From dfa3a41da0d53b6671f7e6611095a39b8a0f2e01 Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Thu, 28 Aug 2025 14:22:37 +0800
Subject: [PATCH 119/123] mtk: mt76: mt7996: fix testmode crash issue

In testmode, the MCU_UNI_EVENT_ALL_STA_INFO event might cause a kernel
crash beacause the monitor interface will be deleted during runtime test.
Therefore, avoid getting all sta info in mac_work when testmode is enabled.

root@OpenWrt:~# [  268.453229] Internal error: Oops: 0000000096000004 [#1] SMP
[  268.458798] Modules linked in: mt7996e(O) mt76_connac_lib(O) mt76(O) mac80211(O) cfg80211(O) compat(O) ksmbd pppoe ppp_async nft_fib_inet l2tp_ppp pptp pppox ppp_mppe ppp_generic nft_reject_ipv6 nft_reject_ipv4 nft_reject_inet nft_reject nft_redir nft_quota nft_numgen nft_nat nft_masq nft_log nft_limit nft_hash nft_fib_ipv6 nft_fib_ipv4 nft_fib nft_ct nft_chain_nat nf_flow_table_netlink nft_flow_offload nf_flow_table_inet nf_tables iptable_nat ipt_REJECT xt_time xt_tcpudp xt_tcpmss xt_statistic xt_state xt_recent xt_quota xt_policy xt_pkttype xt_owner xt_nat xt_multiport xt_mark xt_mac xt_limit xt_length xt_hl xt_helper xt_esp xt_ecn xt_dscp xt_conntrack xt_connmark xt_connlimit xt_connbytes xt_comment xt_cgroup xt_addrtype xt_TCPMSS xt_REDIRECT xt_MASQUERADE xt_LOG xt_HL xt_FLOWOFFLOAD xt_DSCP xt_CLASSIFY ums_usbat ums_sddr55 ums_sddr09 ums_karma ums_jumpshot ums_isd200 ums_freecom ums_datafab ums_cypress ums_alauda spidev slhc sfp rtc_pcf8563 nfnetlink nf_reject_ipv4 nf_nat nf_log_syslog nf_conncount mdio_i2c
[  268.458945]  libcrc32c iptable_mangle iptable_filter ipt_ah ipt_ECN ip_tables br_netfilter at24 mt76qos(O) crypto_safexcel ntfs3 sg i2c_gpio i2c_algo_bit gpio_pca953x i2c_mux_pca954x i2c_mux_pca9541 i2c_mux_gpio i2c_mux ip6table_mangle ip6table_filter ip6_tables ip6t_REJECT x_tables nf_reject_ipv6 nls_ucs2_utils cifs_arc4 asn1_decoder nfsd msdos ip_gre gre l2tp_netlink l2tp_core udp_tunnel ip6_udp_tunnel ipcomp6 xfrm6_tunnel ah6 xfrm4_tunnel ipcomp ah4 tunnel6 tunnel4 ip_tunnel rpcsec_gss_krb5 auth_rpcgss oid_registry xfrm_ipcomp af_key lockd sunrpc grace autofs4 nls_utf8 nls_iso8859_1 nls_cp437 marvell10g marvell crypto_user algif_skcipher algif_rng algif_hash algif_aead af_alg sha512_arm64 sha1_ce cts arc4 uas usb_storage leds_gpio xhci_plat_hcd xhci_pci xhci_mtk_hcd xhci_hcd gpio_button_hotplug(O) vfat fat exfat usbcore usb_common aeon_as21xxx(O) aquantia an8801 [last unloaded: compat(O)]
[  268.627388] CPU: 1 PID: 9327 Comm: napi/phy0-24 Tainted: G           O       6.6.100 #0
[  268.635379] Hardware name: MediaTek MT7988A Reference Board (DT)
[  268.641371] pstate: 80400005 (Nzcv daif +PAN -UAO -TCO -DIT -SSBS BTYPE=--)
[  268.648319] pc : mt7996_mcu_rx_event+0x774/0xfe8 [mt7996e]
[  268.653814] lr : mt7996_mcu_rx_event+0x540/0xfe8 [mt7996e]
[  268.659300] sp : ffffffc0816e3b00
[  268.662603] x29: ffffffc0816e3b00 x28: 0000000000000000 x27: 0000000000000040
[  268.669726] x26: 0000000000000000 x25: 00000080111111e6 x24: ffffff8010193900
[  268.676850] x23: 0000000000000000 x22: ffffff801376802c x21: 0000000000000000
[  268.683972] x20: ffffff8010182900 x19: ffffff8017a3e000 x18: 0000000000000000
[  268.691096] x17: ffffffbfbefa9000 x16: ffffffc080008000 x15: ffffffffffffc0f8
[  268.698218] x14: 0000000000009b98 x13: 000000000000021a x12: 0000000000000002
[  268.705341] x11: 0000000000000000 x10: 00000000000008f0 x9 : 0000000000000000
[  268.712465] x8 : ffffff8017a3e0b4 x7 : 0000000000000000 x6 : 000000000000380f
[  268.719587] x5 : 0000000000000004 x4 : 0000000000000007 x3 : 0000000000000000
[  268.726709] x2 : ffffff8017a3e000 x1 : 0000000000000001 x0 : 0000000000000a77
[  268.733832] Call trace:
[  268.736268]  mt7996_mcu_rx_event+0x774/0xfe8 [mt7996e]
[  268.741406]  mt7996_queue_rx_skb+0xa78/0x13e0 [mt7996e]
[  268.746630]  mt76_dma_rx_poll+0x550/0x6f8 [mt76]
[  268.751244]  __napi_poll+0x34/0x1bc
[  268.754725]  napi_threaded_poll_loop+0x70/0x144
[  268.759243]  napi_threaded_poll+0x70/0x7c
[  268.763241]  kthread+0xd8/0xdc
[  268.766286]  ret_from_fork+0x10/0x20
[  268.769855] Code: 54fff1e8 9129d000 f8607a99 b4fff199 (f941df21)
[  268.775934] ---[ end trace 0000000000000000 ]---
[  268.781019] Kernel panic - not syncing: Oops: Fatal exception in interrupt
[  268.787880] SMP: stopping secondary CPUs
[  268.791795] Kernel Offset: disabled
[  268.795271] CPU features: 0x0,00000000,20000000,1000400b
[  268.800571] Memory Limit: none
[  268.804052] Starting Memory dump SMCC
[  268.807709] Memory dump SMCC failed
[  268.811186] Rebooting in 3 seconds..

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 mt7996/mac.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mt7996/mac.c b/mt7996/mac.c
index 12447f78..ea122f94 100644
--- a/mt7996/mac.c
+++ b/mt7996/mac.c
@@ -3211,7 +3211,7 @@ void mt7996_mac_work(struct work_struct *work)
 	mutex_lock(&mdev->mutex);
 
 	mt76_update_survey(mphy);
-	if (++mphy->mac_work_count % 5 == 0) {
+	if (++mphy->mac_work_count % 5 == 0 && !phy->dev->testmode_enable) {
 		int i;
 
 		mt7996_mac_update_stats(phy);
-- 
2.45.2

